### PR TITLE
Packet number encoding appendix

### LINKED ISSUE #4301: Hard to parse
>  As a result, the size of the packet number encoding is at least one
  bit more than the base-2 logarithm of the number of contiguous
  unacknowledged packet numbers, including the new packet.

That is really hard to parse.

>  For example, if an endpoint has received an acknowledgment for packet
  0xabe8bc, sending a packet with a number of 0xac5c02 requires a
  packet number encoding with 16 bits or more; whereas the 24-bit
  packet number encoding is needed to send a packet with a number of
  0xace8fe.

and so is that.

Indeed I found the remaining paras of the section difficult to follow.

### PR DESCRIPTION
**Pull request by MikeBishop**
Fixes #4301.

I took some liberties with data types here, but... pseudocode.  I do wonder if the slice notation is sufficiently clear; I was hesitant to just do bit-shifts and masks because the number of bytes returned is key.  An alternative would be simply to return the number of bytes, I suppose.

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -7411,6 +7402,43 @@ The initial contents of this registry are shown in {{iana-error-table}}.
...
+  if largest_acked is None:
+    return full_pn
+
+  // The number of bits must be at least one more
+  // than the base-2 logarithm of the number of contiguous
+  // unacknowledged packet numbers, including the new packet.
+  num_unacked = full_pn - largest_acked
+  min_bits = log(num_unacked, 2) + 1
+  num_bytes = ceil(min_bits / 8)
+  return full_pn[-num_bytes:]
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md:7429]: Someone has been doing too much python.

As the value is not encoded into bytes before you index it, this is hard to follow.

```suggestion
  let encoded = []
  for i in 0..num_bytes:
    encoded.unshift((full_pn >> (8 * i)) & 0xff)
    // encoded.unshift((full_pn / (256^i)) % 256) ?
  return encoded
```

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -7411,6 +7402,43 @@ The initial contents of this registry are shown in {{iana-error-table}}.
...
+~~~
+{: #alg-encode-pn title="Sample Packet Number Encoding Algorithm"}
+
+For example, if an endpoint has received an acknowledgment for packet 0xabe8bc
+and is sending a packet with a number of 0xac5c02, there are 29,519 (0x734f)
+outstanding packets.  In order to represent at least twice this range (59,038
+packets, or 0xe69e), 16 bits will be required.
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md:7436]:
```suggestion
packets, or 0xe69e), 16 bits are required.
```

**Review by martinthomson — APPROVED** [review]
I think that this is good, though you need to eliminate a shortcut in the pseudocode.

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -7411,6 +7402,43 @@ The initial contents of this registry are shown in {{iana-error-table}}.
...
+* full_pn is the full packet number of the packet being sent.
+* largest_acked is the largest packet number which has been acknowledged by the
+  recipient in the current packet number space, if any
+
+~~~
+EncodePacketNumber(full_pn, largest_acked):
+  // If no packets in the current space have been
+  // acknowledged, the full packet number is used.
+  if largest_acked is None:
+    return full_pn
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md:7421]: Whoops, missed this.  You can safely assign num_acked to `full_pn + 1` in this case.  Returning early is bad for the same reasons.

Comment (martinthomson) [reply draft-ietf-quic-transport.md:7421]: You can set min_bits to log2(full_pn) (without the +1) as well, I guess.

File: draft-ietf-quic-transport.md [hunk 4]
```diff
@@ -7411,6 +7402,43 @@ The initial contents of this registry are shown in {{iana-error-table}}.
...
 
+# Sample Packet Number Encoding Algorithm {#sample-packet-number-encoding}
+
+The pseudo-code in {{alg-encode-pn}} shows how an implementation can select
+an appropriate size for packet number encodings.
+
+The EncodePacketNumber function takes two arguments:
+
+* full_pn is the full packet number of the packet being sent.
+* largest_acked is the largest packet number which has been acknowledged by the
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:7413]:
```suggestion
* largest_acked is the largest packet number that has been acknowledged by the
```

File: draft-ietf-quic-transport.md [hunk 5]
```diff
@@ -7411,6 +7402,43 @@ The initial contents of this registry are shown in {{iana-error-table}}.
...
+  recipient in the current packet number space, if any
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:7414]:
```suggestion
  peer in the current packet number space, if any
```

File: draft-ietf-quic-transport.md (hunk 4 above)
Comment (janaiyengar) [reply draft-ietf-quic-transport.md:7413]: Can you also modify the variable definitions in the Decode part to match these?

File: draft-ietf-quic-transport.md (hunk 3 above)
Comment (janaiyengar) [reply draft-ietf-quic-transport.md:7421]: @martinthomson : I don't follow what setting num_acked to `full_pn + 1` gets you. Why is this early return not good enough?

File: draft-ietf-quic-transport.md (hunk 1 above)
Comment (janaiyengar) [reply draft-ietf-quic-transport.md:7429]: Someone has been doing too much Javascript :-)
I don't think `.unshift` can be used without definition like this (I had to look it up).

This will be tricky to do without language conventions. I would recommend simply:
```
// return the value after truncating the full packet
// number to num_bytes least significant bytes
return truncate(full_pn, num_bytes)

**Review by janaiyengar — COMMENTED** [review]
This is a great improvement, and I like the symmetry with the decoding algorithm. Thanks, @MikeBishop ! A few comments.

File: draft-ietf-quic-transport.md (hunk 3 above)
Comment (martinthomson) [reply draft-ietf-quic-transport.md:7421]: The early return takes a number and returns it.  We are looking for an encoded value that includes a length (in bytes).  If full_pn is 10, returning that doesn't tell the caller how many bytes to write out.

Running this through the full encoding algorithm avoids that problem.

File: draft-ietf-quic-transport.md (hunk 1 above)
Comment (martinthomson) [reply draft-ietf-quic-transport.md:7429]: So do it this way instead?

```
    encoded = [(either expression), encoded...]
```

File: draft-ietf-quic-transport.md [hunk 6]
```diff
@@ -4382,17 +4382,10 @@ the difference between the largest acknowledged packet and packet number being
 sent.  A peer receiving the packet will then correctly decode the packet number,
 unless the packet is delayed in transit such that it arrives after many
 higher-numbered packets have been received.  An endpoint SHOULD use a large
-enough packet number encoding to allow the packet number to be recovered even
-if the packet arrives after packets that are sent afterwards.
-
-As a result, the size of the packet number encoding is at least one bit more
-than the base-2 logarithm of the number of contiguous unacknowledged packet
-numbers, including the new packet.
```
Comment (kazuho) [inline draft-ietf-quic-transport.md:4390]: I think my preferences goes to retaining this sentence.

Having pseudo code makes sense, but IMO it is better to state the constraints that a function has to meet.

File: draft-ietf-quic-transport.md (hunk 4 above)
Comment (MikeBishop) [reply draft-ietf-quic-transport.md:7413]: There's already a PR for that, which I was emulating.

File: draft-ietf-quic-transport.md (hunk 1 above)
Comment (MikeBishop) [reply draft-ietf-quic-transport.md]: I like the solution of throwing it to an undefined function with a comment describing what it does.  The implementer can do that however their language best permits.

File: draft-ietf-quic-transport.md (hunk 6 above)
Comment (MikeBishop) [reply draft-ietf-quic-transport.md:4390]: It's not actually gone, it's a comment in the pseudo-code.  The paragraph above describes the constraints, while this describes the math which would satisfy those complaints.

File: draft-ietf-quic-transport.md [hunk 7]
```diff
@@ -7411,6 +7402,46 @@ The initial contents of this registry are shown in {{iana-error-table}}.
...
+# Sample Packet Number Encoding Algorithm {#sample-packet-number-encoding}
+
+The pseudo-code in {{alg-encode-pn}} shows how an implementation can select
+an appropriate size for packet number encodings.
+
+The EncodePacketNumber function takes two arguments:
+
+* full_pn is the full packet number of the packet being sent.
+* largest_acked is the largest packet number which has been acknowledged by the
+  peer in the current packet number space, if any
```
Comment (ianswett) [inline draft-ietf-quic-transport.md:7414]:
```suggestion
  peer in the current packet number space, if any.
```

**Review by ianswett — APPROVED** [review]

File: draft-ietf-quic-transport.md (hunk 6 above)
Comment (kazuho) [reply draft-ietf-quic-transport.md:4390]: I think my point is that it would be more helpful to the readers to have this paragraph here, than just having the super-abstract guidance stated in the paragraph above. I tend to think that the specification should be parsable without reading the pseudo-code.

That said, this is a discussion about where the practical guidance should be stated, and therefore my preference is not a strong at all.

File: draft-ietf-quic-transport.md [hunk 8]
```diff
@@ -7411,6 +7402,46 @@ The initial contents of this registry are shown in {{iana-error-table}}.
...
+  if largest_acked is None:
+    num_unacked = full_pn + 1
+  else:
+    num_unacked = full_pn - largest_acked
+
+  min_bits = log(num_unacked, 2) + 1
+  num_bytes = ceil(min_bits / 8)
+
+  // Return the value after truncating the full packet number
+  // to num_bytes least-significant bytes
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md:7431]:
```suggestion
  // to num_bytes least-significant bytes.
```

File: draft-ietf-quic-transport.md [hunk 9]
```diff
@@ -7411,6 +7402,46 @@ The initial contents of this registry are shown in {{iana-error-table}}.
...
+  return truncate(full_pn, num_bytes)
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md:7432]:
```suggestion
  return encode(full_pn, num_bytes)
```

File: draft-ietf-quic-transport.md (hunk 8 above)
Comment (martinthomson) [inline draft-ietf-quic-transport.md:7431]:
```suggestion
  // Encode the integer value and truncating to the 
  // num_bytes least-significant bytes.
```

**Review by janaiyengar — APPROVED** [review]
